### ISSUE TITLE
Coverage is 100% even after bucket distribution is complete. Looks like timeout is happening from node that is down

### ISSUE THREAD

**Issue by 107dipan** [vespa-engine/vespa#22632]
**Describe the bug**
Coverage showing as 100 even after bucket distribution complete after node failure. We wanted to perform some disaster recovery scenarios in vespa and noticed a few discrepencies regarding the coverage value that is returned in response of search. 
We have a cluster of 18 content nodes and have deployed 38 schemas to these content nodes. To see how vespa handles node failure we deleted a k8s pod which was one of our content node. After this we verified that the pending count is 0 from the cluster controller page to ensure bucket distribution is complete. We found thave even after bucket distribution is complete the coverage value we are getting in the response payload is still less than 100. We also tried querying by mentioning the schema but the coverage was still 0. 

My understanding of what is happening is vespa is still trying to send the request to the content node that is down and the request is timing out. This is based on 2 observation, 
1. If we check the response from the 2nd response added the nodes is 646 which is 17*38 and the nodes value is 17 when using one schema. This is sending request to the node that is down. 
2. We are also getting the log line "Connection failure on nodes with distribution-keys: 17" trace level logs

**To Reproduce**
1. Bring up a cluster and feed some data. 
2. Delete one content node from k8s level by deleting the pod. 
3. make search queries once bucket distribution is complete

**Expected behavior**
Coverage should be 100 once bucket distribution is comlete. 

**Environment (please complete the following information):**
 - Infrastructure: Kubernetes
 - Versions [e.g. 22]

**Vespa version**
7.559.12

**Additional context**

**Comment by 107dipan** [comment]
Before disaster - 
{
  "yql": "select * from sources * where status contains \"active\";",
  "timeout": "120s"
}

root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 312398
        },
        "coverage": {
            "coverage": 100,
            "documents": 343082,
            "full": true,
            "nodes": 684,
            "results": 38,
            "resultsFull": 38
        }
		
		
After disaster - 
{
  "yql": "select * from sources * where status contains \"active\";",
  "timeout": "120s"
}

{
    "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 312398
        },
        "coverage": {
            "coverage": 76,
            "documents": 343082,
            "degraded": {
                "match-phase": false,
                "timeout": true,
                "adaptive-timeout": false,
                "non-ideal-state": false
            },
            "full": false,
            "nodes": 646,
            "results": 38,
            "resultsFull": 29
        },
		
		

with one schema - 
{
  "yql": "select * from sources imdgsamclientreportingdocumentslexDocument where status contains \"active\";",
  "timeout": "120s"
}

 "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 3904
        },
        "coverage": {
            "coverage": 0,
            "documents": 5000,
            "degraded": {
                "match-phase": false,
                "timeout": true,
                "adaptive-timeout": false,
                "non-ideal-state": false
            },
            "full": false,
            "nodes": 17,
            "results": 1,
            "resultsFull": 0
        }

**Comment by kkraune** [comment]
I think you should simplify this experiment, as 38 schemas will translate one query to 38 internal queries. Please refer to https://github.com/vespa-engine/sample-apps/tree/master/examples/operations/multinode for a node-down example.

We are not aware of any bug that makes containers send queries to unavailable nodes. Use https://docs.vespa.ai/en/query-api.html#query-tracing to understand the routing of the query to nodes.

**Comment by 107dipan** [comment]
We have already tried making query with one schema and facing the same issue. 

"yql": "select * from sources imdgsamclientreportingdocumentslexDocument where status contains "active";",
"timeout": "120s"
}

"root": {
"id": "toplevel",
"relevance": 1.0,
"fields": {
"totalCount": 3904
},
"coverage": {
"coverage": 0,
"documents": 5000,
"degraded": {
"match-phase": false,
"timeout": true,
"adaptive-timeout": false,
"non-ideal-state": false
},
"full": false,
"nodes": 17,
"results": 1,
"resultsFull": 0
}

**Comment by kkraune** [comment]
In the query trace, do you see that a query is made to the node you set down?

**Comment by 107dipan** [comment]
I can see this message in the query response "Connection failure on nodes with distribution-keys: 17"

**Comment by kkraune** [comment]
In that case, the Vespa stateless container does not consider that node down - it will only forward queries to nodes that are up. It would be interesting to know what is the result of `curl http://localhost:19108/state/v1/health` for the content node you set down (replace localhost:19108 with hostname and port) - refer to https://github.com/vespa-engine/sample-apps/tree/master/examples/operations/multinode-HA

if the content node is down (I.e. you do not get a response), then the problem is the Vespa container that still consider it up.

I am not sure what exactly happens when "Delete one content node from k8s level by deleting the pod." - an alternative test is setting the node down using the state api , see https://docs.vespa.ai/en/content/content-nodes.html#state-api-examples:
 
```
curl -X PUT \
  --data '{"state":{"user":{"state": "down", "reason": "Down a broken node"}}}' \
  http://localhost:19050/cluster/v2/music/storage/0
```

replace localhost with the hostname of a cluster controller

**Comment by 107dipan** [comment]
Actually the reason we were deleting the node from kubernetes and not using this or stop services was because we dont want controlled shut down. But let me try this and see what is happening. 
Since we are deleting the pod altogether we wont be able to use the health url when it is down.

**Comment by kkraune** [comment]
Just to make sure, can you add `ranking.softtimeout.enable=false` to the query as well, to see if the coverage info changes? Thanks

**Comment by 107dipan** [comment]
@kkraune 
I checked if the content node going down was discovered by the clustercontroller and if the containers were getting the updated state of the cluster when a node is going down. 
Here are the logs when the node went down

[2022-05-19 06:22:14.114] INFO    : container        Container.com.yahoo.search.dispatch.searchcluster.Group    Content in group 0, with 18/18 working nodes, is well balanced. Current deviation = 6 %. activeDocs = 921936, skew = 57442, average = 51218
[2022-05-19 07:33:18.885] WARNING : container        Container.com.yahoo.search.cluster.BaseNodeMonitor Taking search node key = 17 hostname = vespa-andromeda-content-17.vespa-andromeda-internal.namespace.svc.cluster.local path = 17 in group 0 statusIsKnown = true working = true activeDocs = 50478 out of service: Connection failure: 10: Backend communication error: Error response from rpc node connection to vespa-andromeda-content-17.vespa-andromeda-internal.namespace.svc.cluster.local:19103: Connection error



[2022-05-19 07:33:18.524] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Failed get node state request: Connection error: Closed at other end. (Node or switch likely shut down)
[2022-05-19 07:33:18.524] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Failed to get node state: D: Connection error: Closed at other end. (Node or switch likely shut down)
[2022-05-19 07:33:18.524] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Stopped or possibly crashed after 7124211 ms, which is before stable state time period. Premature crash count is now 2.
[2022-05-19 07:33:18.524] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Altered min distribution bit count from 58 to 16
[2022-05-19 07:33:18.525] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Node is no longer in slobrok. No pending state request to node.
[2022-05-19 07:33:18.525] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: distributor.17: Altered node state in cluster state from 'U, b 58' to 'D: Connection error: Closed at other end. (Node or switch likely shut down)'
[2022-05-19 07:33:18.525] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Cluster event type SYSTEMSTATE @1652945598525: New cluster state version 35. Change from last: version: 34 => 35, distributor: 17: [Up => Down, minUsedBits: 58 => 16, description:  => Connection error: Closed at other end. (Node or switch likely shut down)]
[2022-05-19 07:33:18.525] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.database.ZooKeeperDatabase     Cluster 'lexdoc': Storing new cluster state version in ZooKeeper: 35
[2022-05-19 07:33:18.527] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.SystemStateBroadcaster Cluster 'lexdoc': Publishing cluster state version 35
[2022-05-19 07:33:18.618] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: storage.17: Failed get node state request: Connection error: Closed at other end. (Node or switch likely shut down)
[2022-05-19 07:33:18.618] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: storage.17: Failed to get node state: D: Connection error: Closed at other end. (Node or switch likely shut down)
[2022-05-19 07:33:18.618] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: storage.17: Stopped or possibly crashed after 4268986 ms, which is before stable state time period. Premature crash count is now 3.
[2022-05-19 07:33:18.619] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: storage.17: Node is no longer in slobrok. No pending state request to node.
[2022-05-19 07:33:18.619] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Added node only event: Event: storage.17: Altered node state in cluster state from 'U' to 'D: Connection error: Closed at other end. (Node or switch likely shut down)'
[2022-05-19 07:33:18.619] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.EventLog       Cluster event type SYSTEMSTATE @1652945598619: New cluster state version 36. Change from last: version: 35 => 36, storage: 17: [Up => Down, description:  => Connection error: Closed at other end. (Node or switch likely shut down)]
[2022-05-19 07:33:18.619] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.database.ZooKeeperDatabase     Cluster 'lexdoc': Storing new cluster state version in ZooKeeper: 36
[2022-05-19 07:33:28.589] INFO    : container-clustercontroller Container.com.yahoo.vespa.clustercontroller.core.SystemStateBroadcaster Cluster 'lexdoc': Publishing cluster state version 36
bash-4.2$

**Comment by 107dipan** [comment]
Also tried querying with "ranking.softtimeout.enable":false and it was still returning with coverage less than 100. 
If the container nodes are getting the state of the content nodes how is the request still going to the content node that is offline.

**Comment by kkraune** [comment]
I suggest you use vespa-visit to dump all the document ids before and after the instance has completed the data migration - just to make sure that the remaining nodes work as they should. 

https://docs.vespa.ai/en/content/visiting.html

**Comment by 107dipan** [comment]
We can observe the same behavior using vespa-stop-services and coverage is down even when distribution is complete. 
bash-4.2$ curl http://vespa-andromeda-content-17.vespa-andromeda-internal.namespace.svc.cluster.local:19107/state/v1/health
curl: (7) Failed connect to vespa-andromeda-content-17.vespa-andromeda-internal.namespace.svc.cluster.local:19107; Connection refused

**Comment by 107dipan** [comment]
Hey @kkraune, 
If the container is sending request to the content nodes that is down even when the cluster controller is detecting node failures and if we dont have anything else to verify from our side we should convert to an issue.

**Comment by kkraune** [comment]
I think Vespa is correctly identifying the node as down, and logging that it cannot connect to it. As it is still in services/hosts.xml, the container will try to use the node once it is up. If it is logged as down , it should not be used in queries

We do not know of any problem that query times out because of nodes being down.

Please use https://docs.vespa.ai/en/query-api.html#query-tracing with trace.timestamps=true to get a detailed dump of the nodes and timing - find a tracelevel that is useful for this purpose. This will identify which node(s) are timing out. No need to put the full output here, it will be huge - but you should be able to identify the content nodes that time out

I don't understand your last comment - this is already an issue

**Comment by 107dipan** [comment]
In our previous conversation you mentioned that the container does not recognize the content node as down. 
"Connection failure on nodes with distribution-keys: 17"

**Comment by kkraune** [comment]
I might have been mistaken. the correct action for the container is to keep retrying as the node is still in the config, but not query it until it is up again. I think the log message is just to signal it cannot connect to it, which is correct

**Comment by 107dipan** [comment]
I checked the traceLevel logs from the query but I can not see any timeout logs in that. If you want I can share the trace level logs. 
Coverage was will down from 100 when redistribution was complete.

**Comment by kkraune** [comment]
OK, so all the 17 nodes nodes responded within timeout, and you got all the expected result then - but still get a timeout in the degraded block?

To make progress, we need a fully reproducible setup with config and data. I don't think we need 18 nodes - something we can run ourselves, maybe built from https://github.com/vespa-engine/sample-apps/tree/master/examples/operations/multinode-HA

**Comment by 107dipan** [comment]
For reproducible steps, we will provide the hosts and services xml and how we are reproducing the issue at our end. 
Do we need to provide the data in the cluster? We should be able to reproduce with any dummy data

**Comment by kkraune** [comment]
as we already have tests for this case, we need data, too, and. We will not have kubernetes environment to run in, though

**Comment by 107dipan** [comment]
Sure @kkraune, 
We will raise a raise with all the steps .

**Comment by 107dipan** [comment]
Hey @geirst , 
Just saw that this was added as a milestone. Does that mean this is a bug? 
Thanks!

**Comment by geirst** [comment]
This issue still has an open dialogue and the milestone was set to signal it is being looked at. No, it doesn't mean it is a bug.

**Comment by 107dipan** [comment]
Let me know in case you need any input from me. 
Thanks!

**Comment by kkraune** [comment]
It seems from the comment in #23412 that this is fixed in Vespa 8.60 - closing
